### PR TITLE
Leuta task6

### PR DESCRIPTION
**Pull request by denisleuta**
Description of the Cartesian Tree Implementation
The CartesianTree class is an implementation of a Cartesian tree, which is a binary tree that maintains the properties of both a binary search tree (BST) and a heap. This class supports operations such as insertion, deletion, and retrieval of key-value pairs, while ensuring that the tree remains balanced according to the priorities assigned to each node.
Key Components
Node Class:
The inner Node class represents a single node in the Cartesian tree. Each node contains:
key: The key associated with the node.
priority: A randomly assigned priority that helps maintain the heap structure.
value: The value associated with the key.
left and right: Pointers to the left and right child nodes.
Initialization:
The CartesianTree constructor initializes an empty tree with a root set to None and a size counter initialized to zero.
Core Methods
Insertion (__setitem__):
The method allows for inserting a new key-value pair into the tree. If the key already exists, its value is updated.
A new node is created with a random priority if the key does not exist. The _split and _merge helper methods are used to maintain the Cartesian tree properties during insertion.
Retrieval (__getitem__):
This method retrieves the value associated with a given key. It raises a KeyError if the key is not found in the tree.
Deletion (__delitem__):
This method deletes a node by its key, using a recursive helper function. If the key is not found, it raises a KeyError. The _merge method is employed to combine left and right subtrees after deletion.
Iteration (__iter__):
Implements in-order traversal of the tree, yielding keys in ascending order.
Reverse Iteration (__reversed__):
Implements reverse in-order traversal, yielding keys in descending order.
Membership Check (__contains__):
Checks if a specific key exists in the tree by attempting to retrieve it.
Size (__len__):
Returns the number of nodes currently stored in the tree.
Helper Methods
_split:
Splits the tree into two subtrees based on a given key. The left subtree contains nodes with keys less than the specified key, while the right subtree contains nodes with keys greater than or equal to that key.
_merge:
Merges two subtrees while preserving Cartesian tree properties based on node priorities. It ensures that higher-priority nodes remain at higher levels in the merged structure.
Summary
The CartesianTree class provides an efficient way to manage ordered data with fast insertion, deletion, and retrieval operations while maintaining balance through random priorities. This implementation leverages Python's built-in features like MutableMapping, allowing it to behave like a dictionary, making it intuitive for users familiar with standard mapping types. Overall, this data structure is particularly useful for applications requiring dynamic sets of ordered elements with efficient access patterns. This description summarizes your code's functionality and structure while highlighting its main components and operations.

### CODE REVIEW

File: tests/test_rgba_decorator.py
Comment (Krekep) [inline tests/test_rgba_decorator.py:1]: Хочу комментарии по поводу удаления этого файла

**Review by Krekep — CHANGES_REQUESTED** [review]
Всё хорошо, но пока не зачёт

File: tests/test_rgba_decorator.py
Comment (denisleuta) [reply tests/test_rgba_decorator.py:1]: Да на самом деле  его удалил, потому что у меня не получается запускать проверку pytest только одного выбранного файла и приходится запускать проверку всех файлов. А файл test_rgba_decorator проверяется очень долго и я его убрал и забыл вернуть. Сейчас добавил его заново. В следующий раз разберусь, как проверять только нужный файл, чтобы ничего не удалять

Comment (Krekep) [reply tests/test_rgba_decorator.py:1]: Лучше просто не проверять последнюю комбинацию, а остановится на какой-нибудь тысячной

**Review by Krekep — APPROVED** [review]
Зачтено

**Pull request merged**